url64d634b8f0_agdturner_generic_tgz-pJ8-uk_ac_leeds_ccg_andyt_generic_visualisation_charts_Generic_BarChartJ8/src/uk/ac/leeds/ccg/andyt/generic/math/Generic_float.java

```diff
@@ -0,0 +1,325 @@
+//     */
+//    public static float next(
+//            float value) {
+//        if (value == Float.MAX_VALUE) {
+//            System.out.println(
+//                    "Warning: Returning Float.POSITIVE_INFINITY in "
+//                    + Generic_float.class + ".next(" + value + ")");
+//            return Float.POSITIVE_INFINITY;
+//        } else {
+//            float MIN_VALUE = Float.MIN_VALUE;
+//            float MIN_NORMAL = Float.MIN_NORMAL;
+//            float difference = MIN_NORMAL;
+//            //float difference = MIN_VALUE;
+//            float previousDifference = difference;
+//            boolean calculated = false;
+//            while (!calculated) {
+//                if ((value + difference) == value) {
+//                    previousDifference = difference;
+//                    System.out.println(difference);
+//                    difference += MIN_NORMAL;
+//                    //difference += MIN_VALUE;
+//                } else {
+//                    calculated = true;
+//                }
+//            }
+//            return value + previousDifference;
+//        }
+//    }
+
+    /**
+     * @param a
+     * @param b
+     * @return a + b > a;
+     */
+    public static boolean isAddSignificant(float a, float b) {
+        return a + b > a;
+    }
+
+    /**
+     * @param a
+     * @param b
+     * @return a - b < a;
+     */
+    public static boolean isSubtractSignificant(float a, float b) {
+        return a - b < a;
+    }
+
+    public static float next(
+            float value,
+            float lowerbound,
+            float upperbound) {
+        float halfway = (upperbound - lowerbound) / 2f;
+        if (isAddSignificant(value, halfway)) {
+            return next(value, lowerbound, lowerbound + halfway);
+        } else {
+            if (halfway > 0f) {
+                return next(value, lowerbound + halfway, upperbound);
+            } else {
+                return value + lowerbound;
+            }
+        }
+    }
+
+    public static float next(
+            float value) {
+        if (value == Float.MAX_VALUE) {
+            System.out.println(
+                    "Warning: Returning Float.POSITIVE_INFINITY in "
+                    + Generic_float.class + ".next(" + value + ")");
+            return Float.POSITIVE_INFINITY;
+        } else {
+            float MAX_VALUE = Float.MAX_VALUE;
+            float diff_max = value - MAX_VALUE;
+            float MIN_VALUE = Float.MIN_VALUE;
+            if (diff_max == MIN_VALUE) {
+                // Would this ever happen?
+                return MAX_VALUE;
+            }
+            float one = 1f;
+            float two = 2f;
+            float diff2 = one;
+            while (isAddSignificant(value, diff2)) {
+                diff2 /= two;
+            }
+            // diff2 < next <= (diff2 * two)
+            return next(value, diff2, diff2 * two);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 248:269, 274:295.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.